### ISSUE TITLE
Crash when calling `jl_load` with file with syntax errors.

### ISSUE THREAD

**Issue by twadleigh** [JuliaLang/julia#9507]
I am embedding julia. My embedding seems to work fine as long as I call `jl_load` on a syntactically correct file. However, if the file has syntax errors, it brings the entire process down, and I see this output:

`fatal: error thrown and no exception handler available.`

with no opportunity to check `jl_exception_occurred`. Is this a bug? Or is there a way to register an exception handler when embedding, in which case it would just be a deficiency in the docs?

**Comment by tknopp** [comment]
This is really a little tricky because of the use of C exceptions. @twadleigh the jlapi.c file is kind of a separate module specially designed for the embedding use case. Thats why `jl_eval_string` worked as you expected.

I think the most important question is weather we need a public C API that uses exceptions or not. @JeffBezanson: Opinions on that?

Given the recent discussions on code quality we may should make very clear that the C API is experimental

**Comment by JeffBezanson** [comment]
I think this particular case is covered by calling `jl_call1` with the `include` function, or using `jl_eval_string` as above.

Everything in `jlapi.c` should have clearly-defined exception behavior, and I believe every function there currently catches exceptions. If there are things you can't reasonably do with `jl_call` and `jl_eval_string` and which throw exceptions, we should add wrappers for them to jlapi.c as needed. I don't think it's worth the time to make exception/non-exception versions of every function up front.

**Comment by tknopp** [comment]
Jeff, I agree with you. The embedding API should only be a handful functions that are necessary so my intention was not to make for every function an exception free version. Still I think we can improve the current situation by making more clear what these "official" embedding functions are, no?

The question is if this is only a documentation issue or if the "public" api function should be somehow highlighted (e.g. by a common `jl_api_...` prefix or by putting them into a dedicated header `julia_api.h`)

**Comment by twadleigh** [comment]
I'm closing this as straight-up user error. I went to make a PR to suggest a few changes to the embedding documentation, particularly the removal or adding sufficient caveats to the use of `jl_load`, only to find that `jl_load` is not mentioned there. Nor is it mentioned in its history, as far as I can tell. I would have sworn I was following the the embedding docs when I chose to use `jl_load` in my application.

Sorry for the noise.
